### PR TITLE
Implement journeys anticipation

### PR DESCRIPTION
**Pull request by nberard**
request (1.1.5) before : 

```
SELECT
    ABS(requested_date_time::date - request_date::date) as ecart,
    count(1)
FROM
    stat_compiled.journey_request_stats jrs
WHERE
    region_id = 'fr-auv'
    AND request_date >= :date_begin 
    AND request_date < (:date_end :: date) + interval '1 day'
    AND is_internal_call = 0
GROUP BY 1
ORDER BY 1
```
time: 1min38s
request after : 

```
SELECT difference, nb
FROM
    stat_compiled.coverage_journey_anticipations 
WHERE
    region_id = 'fr-auv'
    AND request_date >= '2017-02-17'
    AND request_date < ('2017-02-17' :: date) + interval '1 day'
    AND is_internal_call = 0
ORDER BY 1
```
time: less than a second

### CODE REVIEW

File: migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import config
+from migrations.utils import get_create_partition_sql_func, get_drop_partition_sql_func, \
+                             get_create_trigger_sql
+
+table = "coverage_journey_anticipations"
+schema = config.db['schema']
+
+
+def upgrade():
+    op.create_table('coverage_journey_anticipations',
```
Comment (azime) [inline migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py]: table

File: migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+                    sa.Column('region_id', sa.Text(), nullable=False),
+                    sa.Column('is_internal_call', sa.SmallInteger(), nullable=False),
+                    sa.Column('request_date', sa.DateTime(), nullable=True),
+                    sa.Column('difference', sa.Integer(), nullable=True),
+                    sa.Column('nb', sa.BigInteger(), nullable=True),
+                    sa.PrimaryKeyConstraint('region_id', 'is_internal_call'),
+                    sa.UniqueConstraint('region_id', 'is_internal_call', 'request_date',
+                                        name='{schema}_coverage_journey_anticipations_pkey'.format(
+                                            schema=schema)),
```
Comment (azime) [inline migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py]: '{schema}_{table}_pkey'.format(
                                            schema=schema, table=table))

File: models.py [hunk 3]
```diff
@@ -8,14 +8,14 @@
...
-    start_network_id = Column(Text(), primary_key=True,  nullable=False)
-    start_network_name = Column(Text(), primary_key=False,  nullable=False)
-    end_network_id = Column(Text(), primary_key=True,  nullable=False)
-    end_network_name = Column(Text(), primary_key=False,  nullable=False)
-    request_date = Column(DateTime(), primary_key=True,  nullable=False)
-    is_internal_call = Column(SmallInteger(), primary_key=True,  nullable=False)
-    nb = Column(BigInteger(), primary_key=False,  nullable=False)
+    region_id = Column(Text(), primary_key=True, nullable=False)
+    start_network_id = Column(Text(), primary_key=True, nullable=False)
+    start_network_name = Column(Text(), primary_key=False, nullable=False)
```
Comment (azime) [inline models.py]: nullable=False

File: models.py [hunk 4]
```diff
@@ -8,14 +8,14 @@
...
+    end_network_id = Column(Text(), primary_key=True, nullable=False)
+    end_network_name = Column(Text(), primary_key=False, nullable=False)
```
Comment (azime) [inline models.py]: nullable=False

File: models.py [hunk 5]
```diff
@@ -224,3 +212,30 @@ class CoverageLines(Base):
...
+        ),
+    )
+
+
+class CoverageAnticipation(Base):
+    __tablename__ = 'coverage_journey_anticipations'
+
+    region_id = Column(Text(), primary_key=True, nullable=False)
+    is_internal_call = Column(SmallInteger(), primary_key=True, nullable=False)
+    request_date = Column(DateTime(), primary_key=False, nullable=True)
```
Comment (vincentlepot) [inline models.py]: request_date is nullable ??

Comment (vincentlepot) [inline models.py]: If not part of primary key, you will have trouble :)

File: models.py [hunk 6]
```diff
@@ -224,3 +212,30 @@ class CoverageLines(Base):
...
+    difference = Column(Integer(), primary_key=False)
+    nb = Column(BigInteger(), primary_key=False)
+    __table_args__ = (
+        UniqueConstraint(
+            'region_id', 'is_internal_call', 'request_date', name='coverage_journey_anticipations_pkey'
```
Comment (vincentlepot) [inline models.py]: If difference is not in unique constraint, you will have trouble too...

File: models.py [hunk 7]
```diff
@@ -224,3 +212,30 @@ class CoverageLines(Base):
...
+    )
+
+
+class CoverageAnticipation(Base):
+    __tablename__ = 'coverage_journey_anticipations'
+
+    region_id = Column(Text(), primary_key=True, nullable=False)
+    is_internal_call = Column(SmallInteger(), primary_key=True, nullable=False)
+    request_date = Column(DateTime(), primary_key=False, nullable=True)
+    difference = Column(Integer(), primary_key=False)
```
Comment (vincentlepot) [inline models.py]: If not part of primary key, you will have trouble :)

File: models.py (hunk 6 above)
Comment (vincentlepot) [inline models.py]: I don't see the need for a unique constraint here in addition to the primary key

File: migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+
+def upgrade():
+    op.create_table('coverage_journey_anticipations',
+                    sa.Column('region_id', sa.Text(), nullable=False),
+                    sa.Column('is_internal_call', sa.SmallInteger(), nullable=False),
+                    sa.Column('request_date', sa.DateTime(), nullable=True),
+                    sa.Column('difference', sa.Integer(), nullable=True),
+                    sa.Column('nb', sa.BigInteger(), nullable=True),
+                    sa.PrimaryKeyConstraint('region_id', 'is_internal_call'),
```
Comment (vincentlepot) [inline migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py]: Are you sure this is actually your primary key?

File: models.py (hunk 5 above)
Comment (nberard) [inline models.py]: thanks ;)

File: migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py (hunk 1 above)
Comment (nberard) [inline migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py]: done

File: migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py (hunk 2 above)
Comment (nberard) [inline migrations/alembic/versions/1c26fbe7af9_coverage_journey_anticipations.py]: done

File: models.py (hunk 3 above)
Comment (nberard) [inline models.py]: can't see it

File: models.py (hunk 4 above)
Comment (nberard) [inline models.py]: can't see it

File: models.py (hunk 3 above)
Comment (azime) [inline models.py]:
````python
start_network_name = Column(Text(), primary_key=False, nullable=True)
````

**Comment by vincentlepot** [comment]
Tests have not been run by Jenkins automatically?

**Comment by vincentlepot** [comment]
Jenkins test this please

**Comment by datanel** [comment]
you can rebase and add an integration test :wink:

**Pull request merged**